clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashCollection.java

```diff
@@ -127,35 +132,136 @@ private static void removeFromList(Element head, Element[] elements, int element
         element.setPrev(INVALID_INDEX);
     }
 
-    private class ImplicitLinkedHashSetIterator implements Iterator<E> {
+    private class ImplicitLinkedHashCollectionIterator implements ListIterator<E> {
+        private int cursor = 0;
         private Element cur = head;
 
-        private Element next = indexToElement(head, elements, head.next());
+        // private Element next = indexToElement(head, elements, head.next());
+        private Element lastReturned;
 
         @Override
         public boolean hasNext() {
-            return next != head;
+            return cursor != size;
+        }
+
+        @Override
+        public boolean hasPrevious() {
+            return cursor != 0;
         }
 
         @Override
         public E next() {
-            if (next == head) {
+            if (cursor == size) {
+                throw new NoSuchElementException();
+            }
+            cur = indexToElement(head, elements, cur.next());
+            ++cursor;
+            @SuppressWarnings("unchecked")
+            E returnValue = (E) cur;
+            lastReturned = returnValue;
+            return returnValue;
+        }
+
+        @Override
+        public E previous() {
+            if (cursor == 0) {
                 throw new NoSuchElementException();
             }
-            cur = next;
-            next = indexToElement(head, elements, cur.next());
             @SuppressWarnings("unchecked")
             E returnValue = (E) cur;
+            lastReturned = returnValue;
+            cur = indexToElement(head, elements, cur.prev());
+            --cursor;
             return returnValue;
         }
 
+        @Override
+        public int nextIndex() {
+            return cursor;
+        }
+
+        @Override
+        public int previousIndex() {
+            return cursor - 1;
+        }
+
         @Override
         public void remove() {
-            if (cur == head) {
+            if (lastReturned == null) {
                 throw new IllegalStateException();
             }
-            ImplicitLinkedHashSet.this.remove(cur);
-            cur = head;
+
+            if (lastReturned == cur) {
+                cursor--;
+                cur = indexToElement(head, elements, lastReturned.prev());
+            }
+            ImplicitLinkedHashCollection.this.remove(lastReturned);
```

Comment:
This does work, but for a reason that is a bit obscure.  When using an `ImplicitLinkedHashMultiCollection`, `remove` will remove the element b such that a == b, if it exists.  This is necessary since if it just took the first element where `a.equals(b)`, it might be a different one than expected.

It might be clearer to directly call `removeElementAtSlot`, since we already know the slot number.

Reply:
Good call, that's better. Done